connector/src/main/java/com/google/cloud/hive/bigquery/connector/input/BigQueryInputSplit.java

```diff
@@ -0,0 +1,211 @@
+      // "hive.io.file.readcolumn.names" when multiple tables are read in the
+      // same query. So we have to select all the columns (i.e. `SELECT *`).
+      // This is unfortunately quite inefficient. Tez, however, does not have that issue.
+      // See more info here: https://lists.apache.org/thread/g464zybq4g6c7p2h6nd9jmmznq472785
+      // TODO: Investigate to see if we can come up with a workaround. Maybe try
+      //  using the new MapRed API (org.apache.hadoop.mapreduce) instead of the old
+      //  one (org.apache.hadoop.mapred)?
+      selectedFields = new HashSet<>(columnNames);
+    } else {
+      selectedFields =
+          new HashSet<>(Arrays.asList(ColumnProjectionUtils.getReadColumnNames(jobConf)));
+    }
+
+    // If a WHERE clause with filters is present, translate the filter values to
+    // be compatible with BigQuery
+    String serializedFilterExpr = jobConf.get(TableScanDesc.FILTER_EXPR_CONF_STR);
+    ExprNodeGenericFuncDesc filterExpr;
+    Optional<String> filter = Optional.empty();
+    if (serializedFilterExpr != null) {
+      filterExpr = SerializationUtilities.deserializeExpression(serializedFilterExpr);
+      ExprNodeGenericFuncDesc translatedFilterExpr =
+          (ExprNodeGenericFuncDesc) BigQueryFilters.translateFilters(filterExpr);
+      filter = Optional.of(translatedFilterExpr.getExprString());
+    }
+
+    // TODO: If the BigQuery doesn't exist, then readSessionCreator.create() throws
+    //  a NullPointerException. This is because the ReadSessionCreator class in the
+    //  bigquery-connector-common library does not set the "setThrowNotFound" option
+    //  when it calls the BQ API's getTable() method. We should maybe modify that
+    //  library's code to better handle this case and provide a better error message
+    //  instead of just throwing a NullPointerException.
+    //  See: https://github.com/GoogleCloudDataproc/spark-bigquery-connector/issues/640
+    ReadSessionResponse readSessionResponse =
+        readSessionCreator.create(opts.getTableId(), ImmutableList.copyOf(selectedFields), filter);
+    ReadSession readSession = readSessionResponse.getReadSession();
+    Path warehouseLocation = new Path(jobConf.get("location"));
+    FileSplit[] fileSplits = new FileSplit[readSession.getStreamsCount()];
+    int i = 0;
+    for (ReadStream readStream : readSession.getStreamsList()) {
+      fileSplits[i++] =
+          new BigQueryInputSplit(warehouseLocation, readStream.getName(), columnNames);
+    }
+    return fileSplits;
+  }
+
+  /**
+   * Creates and returns a ReadRowsHelper from the bigquery-connector-common library. The helper
+   * takes care of reading data from the split's BQ stream.
+   */
+  public ReadRowsHelper getReadRowsHelper(Configuration conf) {
+    if (readRowsHelper == null) {
+      ReadRowsRequest.Builder request =
+          ReadRowsRequest.newBuilder().setReadStream(checkNotNull(getStreamName(), "name"));
+      Injector injector =
```

Comment:
Ok, I'll look into it

Reply:
Yes agreed. I think I'm probably not using the injector the right way. This can be improved later as part of https://github.com/GoogleCloudDataproc/hive-bigquery-connector/issues/22